### ISSUE TITLE
[New-Model] Implement Multilingual Speech Synthesis

### ISSUE THREAD

**Issue by WeberJulian** [mozilla/TTS#599]
I was wondering if anyone else would be interested by the implementation of this paper in the mozilla/TTS repo : 
"[Learning to Speak Fluently in a Foreign Language: Multilingual Speech Synthesis and Cross-Language Voice Cloning](https://google.github.io/tacotron/publications/multilingual/index.html)"

I think that having the possibility of using code-switching is a huge plus for non English models since English is use in everyday life and not being able to pronounce English words in French for example limit the usability of the model. (my model can't say parking)

Furthermore, I hope that combining this with the new encoder we have trained would maybe allow for voice cloning in language with low resources (or at least have more voices available).

I'm a beginner when it comes to pytorch but I would love to help implementing this paper although I'm not sure I can do it  alone.

What do you think ? Would it be interesting to have that in the repo ? would it be hard to implement ? Who would be willing to help ?

Thanks for reading

**Comment by WeberJulian** [comment]
Sorry, I was busy on other projects but I have more time to work on this now. Have you read the paper @erogol ? Are you still interested ? If not, @Edresson maybe I can help on the similar thing that you are working on ?

**Comment by Edresson** [comment]
Hi @WeberJulian ,

Sorry everything was very busy and I couldn't see it.

I read the article and really interesting. In addition to this article I would like to list the article [End-to-End Code-Switched TTS with Mix of Monolingual Recordings ](https://csttsdemo.github.io/).

In addition, the article: [One Model, Many Languages: Meta-learning for Multilingual Text-to-Speech](https://arxiv.org/pdf/2008.00768.pdf) combines the best of both approaches.

I believe that the best way is to implement this last article.

On this demo page you have the comparison of the three models trained under the same conditions: https://tomiinek.github.io/multilingual_speech_samples/

In addition, the [implementation](https://github.com/Tomiinek/Multilingual_Text_to_Speech) is open source (with the MIT license) which we can use as a basis for implementation in TTS (as done in GlowTTS).

I can do my best to help you with the implementation as it is very interesting for the community. In addition, we can try to propose some improvements to the method.

**Comment by WeberJulian** [comment]
Sounds great ! I'm gonna take some time to read those articles/papers you shared and I'll get back to you.

**Comment by WeberJulian** [comment]
@Edresson I listened the sample from the [demo page](https://tomiinek.github.io/multilingual_speech_samples/) you shared and indeed the generated approach compares favorably to the fully shared and separated approaches. What bothers me is that [the samples](https://google.github.io/tacotron/publications/multilingual/index.html) from the authors of "Multilingual speech synthesis and cross-language voice cloning" sounds so much better than the one from the demo page that compares the approach.
What do you think explain the disparity ?
The team from google has a much better trained vocoder ?
They spend more time tunning the hyperparameters ?
Higher quality dataset ? (although they only use 3 ES and 5 CN speakers )
They cherry picked the samples ?
Bad implementation from the author of the demo page ?

**Comment by WeberJulian** [comment]
But yeah the samples from the generated approach sound good to me, I'm gonna study the code and the paper to be sure that I understand how it works and how it could be implemented here.

**Comment by Edresson** [comment]
> 
> 
> @Edresson I listened the sample from the [demo page](https://tomiinek.github.io/multilingual_speech_samples/) you shared and indeed the generated approach compares favorably to the fully shared and separated approaches. What bothers me is that [the samples](https://google.github.io/tacotron/publications/multilingual/index.html) from the authors of "Multilingual speech synthesis and cross-language voice cloning" sounds so much better than the one from the demo page that compares the approach.
> What do you think explain the disparity ?
> The team from google has a much better trained vocoder ?
> They spend more time tunning the hyperparameters ?
> Higher quality/length dataset ? (although google only use 3 ES and 5 CN speakers which is impressive) Common voice (used by generated) can be very noisy
> They cherry picked the samples ?
> Bad implementation from the author of the demo page ?

I believe it can be a mix of the various factors that you mentioned. The datasets used by Google should be of better quality. Hyperparameters can also be a problem. The interesting thing is that the comparison between the models is done in the same scenario (dataset, hyperparameters and the like), so there is a fairer comparison.

**Comment by WeberJulian** [comment]
I've spent some time looking at the code and I don't know what a good first step would be ^^' I think I'm gonna try training a model with the original repo and mailabs to test it out.

**Comment by WeberJulian** [comment]
In the paper, they say 

> To train code-switching models, multi-speaker data is required to disentangle the connection between languages and speakers

and they use it as a reason to use common voice in their dataset. I was wondering if instead of the "speaker classifier block" they use, we could use @mueller91's pretrained encoder. That way, we don't have to use common voice to increase speaker count and thus lower the quality of the generated voice. Do you think it could work/be useful @Edresson ?

**Comment by Edresson** [comment]
> In the paper, they say
> 
> > To train code-switching models, multi-speaker data is required to disentangle the connection between languages and speakers
> 
> and they use it as a reason to use common voice in their dataset. I was wondering if instead of the "speaker classifier block" they use, we could use @mueller91's pretrained encoder. That way, we don't have to use common voice to increase speaker count and thus lower the quality of the generated voice. Do you think it could work/be useful @Edresson ?
> 
> EDIT: nevermind I think the classifier is important for the trainning of the encoder...
> 
> > The classifier follows principles of domain adversarial training [22] and is used to proactively remove speaker-specific information from the encoders.

I had thought about the future of using the speaker encoder instead of the speaker classifier block. However, as you mentioned, it is apparently important for encoder training. In the future we can think of some way to integrate, using the speaker encoder would bring many advantages.

**Comment by WeberJulian** [comment]
We could maybe let the speaker classifier block here but only concatenate in the decoder's input the encoding generated by our pretrained encoder like so

![image](https://user-images.githubusercontent.com/17219561/106642068-34291f80-6588-11eb-8037-d7c620fcfeb4.png)

Do you think it could work ?

**Comment by erogol** [comment]
what does this gradient reversal do? (just looking at the figure)

**Comment by WeberJulian** [comment]
Sorry it took a while but I wanted to make sure I understood it well before trying to explain it to you.

I found [this great article](https://medium.com/@2017csm1006/unsupervised-domain-adaptation-by-backpropagation-da730a190fd2) that explains domain adversarial training but here is the TL;DR :

> Basically, the feature extractor will try to perform some transformation on the source and target instances such that the transformed instances appear as if it is coming from the same distribution and domain classifier will not be able to classify the domain of the transformed instances. This is achieved by training both the feature extractor and domain classifier in such a way that feature extractor will be trained to maximize the domain classification loss, while domain classifier will try to minimize the domain classification loss. So, this is similar to adversarial training wherein the feature extractor is trying to confuse domain classifier by bringing the two distributions closer. For the transformed source instances the label predictor will be trained for predicting the labels of the source instances. The feature extractor will therefore be trained to minimize classification loss of the label predictor and maximize classification loss of domain predictor. the label predictor and domain predictor will be trained to minimize there respective classification loss.

And this is the explanation for the gradient reversal layer:

> The Gradient Reversal Layer basically acts as an identity function (outputs is same as input) during forward propagation but during back propagation it multiplies its input by -1. Intuitively, During back propagation the output of GRL is basically leading to the opposite of Gradient descent that is performing gradient ascent on the feature extractor with respect to the classification loss of Domain predictor.

So here this is used to remove any information about the speaker in the encoder. That's useful I think because intuitively the encoder should be affected by the language but not the speaker and that way, we disentangle the language representation (added before the encoder) from the speaker representation (which is added later in the decoder).

I hope I answered your question @erogol !

**Comment by erogol** [comment]
thx for the explanation :)

**Comment by erogol** [comment]
After finally reading the paper here are the things I see we need to do to get it implemented.
- Add Language embedding layer to Tacotron models
- Add Variational AutoEncoder as an alternative to GST
- Speaker classifier with Gradient Reversal layer.
- Enable UTF-8 inputs to TTS. (They report that it does not work actually.)
- Incorporate tone information for tonal languages like Mandarin.

Implementing the first 3 steps would at least give us the chance to try Latin alphabet models.

**Comment by WeberJulian** [comment]
We also need to add a way to create those kind of batches (I don't know if it's necessary)
![image](https://user-images.githubusercontent.com/17219561/107067169-30430a80-67df-11eb-8151-5425733dcdff.png)

**Comment by erogol** [comment]
appended to the list :)

**Comment by WeberJulian** [comment]
And the parameter generator (for the encoder) also seems to be a crucial part of why this architecture works for multiple languages

**Comment by erogol** [comment]
```the parameter generator``` what you mean by that?

**Comment by WeberJulian** [comment]
The green part
![image](https://user-images.githubusercontent.com/17219561/107070567-d133c480-67e3-11eb-82cb-91ef7ebf92e4.png)

**Comment by erogol** [comment]
wait did I read the wrong paper. https://arxiv.org/pdf/1907.04448.pdf

Where is your figure from?

**Comment by WeberJulian** [comment]
Haha I was starting to supect that you read the wrong paper. 

@Edresson thaugh that this paper was more intresting than the one I shared because it combined the best of the two approches. 

> In addition, the article: [One Model, Many Languages: Meta-learning for Multilingual Text-to-Speech](https://arxiv.org/pdf/2008.00768.pdf) combines the best of both approaches.
> 
> I believe that the best way is to implement this last article.

**Comment by erogol** [comment]
Ive not read it but why do you think it is more interesting?

**Comment by WeberJulian** [comment]
That's comment 4 to 8 on this issue :)

**Comment by erogol** [comment]
yeah I've read but yet I can't judge why one is better than another. I guess I wait until I read the other paper.

**Comment by WeberJulian** [comment]
Well I find the samples of the paper from google better. That said, when trained and compared on the same dataset https://tomiinek.github.io/multilingual_speech_samples/ I have to admit that the generated approaches does better than the shared one. In the paper you'll find comparasion as well that show that the generated approach needs less data do learn as well.
![image](https://user-images.githubusercontent.com/17219561/107116485-cd4f8300-6873-11eb-89c5-e40e7f05a121.png)
SLG = separated
SHA = shared (paper from google)
GEN = generated

**Comment by WeberJulian** [comment]
I just realised that the authors of "One Model, Many Languages: Meta-learning for Multilingual Text-to-Speech" didn't mention anything about phonemes in their implementation of the shared model. And they use graphmes to train their generated model. That might be why the shared approach did so badly in their comparaison since the paper from google said that phonemes played a role in the ability to share parameters between languages.

(paper from google)
![image](https://user-images.githubusercontent.com/17219561/107116702-5915df00-6875-11eb-865f-18b8781a7800.png)

**Comment by WeberJulian** [comment]
And you @erogol, do you think that could be why the samples from [here](https://tomiinek.github.io/multilingual_speech_samples/) sound so much worse than in Google's implementation ? That would also  explain why the generated model compares favorably in low ressource training to the shared approach in tomiinek's paper.

**Comment by WeberJulian** [comment]
Hi, are you guys still interested ? What should we do ?

**Comment by erogol** [comment]
Yep I'm still interested and want to help. I dont think I can take the lead but  the rest is assured.

Sorry I was a bit slow in reponding. I'll check the second paper in detail asap and get back to you.

**Comment by erogol** [comment]
So my understanding is that the main difference is the parameter generator btw these two paper.  Would you agree? 

After the first look to the 2nd paper, they don't look too different. And it is hard to judge fairly since they don't use the same dataset.

However, google's work looks easier to implement and since they are not too different we can introduce the parameter generator of the second paper later on. 

I also agree that not using phonemes in the 2nd work might be the reason of the perf gap between the baselines and the proposed model.

**Comment by Tomiinek** [comment]
**Hey guys, thanks @erogol  for reaching me!**

I am the author of the _"One Model, Many Languages: Meta-learning for Multilingual Text-to-Speech"_ paper and I would join your discussion if you do not mind.


### I went through the issue and I've got a few notes:

> I've spent some time looking at the code and I don't know what a good first step would be ^^' I think I'm gonna try training a model with the original repo and mailabs to test it out.

I had hard times using the mailabs dataset. Even though it is large and includes many languages, the source of the data differs from language to language. That is why I used CSS10 data at the end, it is much more cleaner. It would be nice to combine it with the TUNDRA dataset, which is segmented in a similar fashion and adds other languages such as English, Danish etc.

> @Edresson I listened to the samples from the demo page you shared and indeed the generated approach compares favorably to the fully shared and separated approaches. What bothers me is that the samples from the authors of "Multilingual speech synthesis and cross-language voice cloning" sounds so much better than the one from the demo page that compares the approach.

I think it is mainly because of the fact that they (1) used phnemes, (2) have much more better data, and (3) have a perfect vocoder.

> Higher quality/length dataset ? (although google only use 3 ES and 5 CN speakers which is impressive) Common voice (used by generated) can be very noisy

I used a very small and clean subset of the Common voice because I needed more speakers to make the code-switching working.  It would be nice to have LJ Speech and VCTK for all the langauges :smile: 

> Bad implementation from the author of the demo page ?

I hope this is not the problem :innocent: 

> and they use it as a reason to use common voice in their dataset. I was wondering if instead of the "speaker classifier block" they use, we could use @mueller91's pretrained encoder. That way, we don't have to use common voice to increase speaker count and thus lower the quality of the generated voice. Do you think it could work/be useful @Edresson ?
> EDIT: nevermind I think the classifier is important for the trainning of the encoder...

The speaker classifier helps a bit to disentangle the speaker and language.  I think it is much more important when training on phoneme inputs. But it is not easy to tune its parameters, it can diverge or result in no effect easily.

> We also need to add a way to create those kind of batches (I don't know if it's necessary)

I think you want it (see [these lines](https://github.com/Tomiinek/Multilingual_Text_to_Speech/blob/master/utils/samplers.py#L50-L122)). The model should not have a bias towards a language and moreover, it is the most straightforward way of implementing the generated encoder.

> I just realised that the authors of "One Model, Many Languages: Meta-learning for Multilingual Text-to-Speech" didn't mention anything about phonemes in their implementation of the shared model. And they use graphmes to train their generated model. That might be why the shared approach did so badly in their comparaison since the paper from google said that phonemes played a role in the ability to share parameters between languages.

You are right. When using graphemes, the advantage of the generated encoder is more emphasized, because it is able to share even some complex low-level pronunciation rules across languages and it helps a lot when training in a low-resource setting.

I believe that the usage of phonemicized inputs improves the output quality including the code-switching ... however, you need to be able to phonemicize all the languages into the same phoneme alphabet. And do not forget about the tones :disappointed:  I tried to used IPA as the common alphabet and Epitran for the phonemicization itself,  [see this function](https://github.com/Tomiinek/Multilingual_Text_to_Speech/blob/master/utils/text.py#L35-L96). At the end, I used graphemes to emphasize the sharing capabilities of the generated encoder.

I also believe that the usage of phonemes would reduce the requirement for having multiple speakers per language. 

> However, google's work looks easier to implement and since they are not too different we can introduce the parameter generator of the second paper later on.

See [these lines](https://github.com/Tomiinek/Multilingual_Text_to_Speech/blob/master/modules/encoder.py#L159-L221) for my implementation of the encoder.


Hope it helps :slightly_smiling_face:

**Comment by WeberJulian** [comment]
Thanks a lot @Tomiinek for taking the time to clarify those points !

I think having an implementation of both models could be really interesting since the paper from Google seem to produce really good results where a phonemizer is available and when the data quality is great. But implementing your model could enable us to train TTS systems for small dialects for which there are no phonemizers and a small hand-made dataset. (since your model generalize so well on graphemes in low data senarios).

In any case, thanks for your article and especially for publishing your code on github (it's very useful for beginners like me)

**Comment by erogol** [comment]
As far as I can see there is no official implementation of Google's. It is maybe better to start with Tomiinke's work as we can rely on his implementation.

**Comment by erogol** [comment]
How about this list to follow;

- Decide the target dataset. (perhaps from the paper to compare results)
- Transition Tomiinek's implementation to TTS.
- Train and compare with the original implementation.
- Then start to wander possible improvements 🚀

**Comment by WeberJulian** [comment]
Yeah I agree.
So how should we go about this ? 
Open a new branch on this repo ? 
Or should I fork, and create a branch from master ? dev ?

**Comment by erogol** [comment]
I created the branch "multilingual". So we can start working on it sending PRs to that branch.

**Comment by WeberJulian** [comment]
Since I'm a bit new to opensource and complex project like TTS I'd like to divide the work in parts that we can test by training a functional model afterward. 
I was thinking maybe a first step would be to implement the speaker classifier and gradient reversal layer and train it on VCTK to see how well it works ?
What do you think ?

**Comment by erogol** [comment]
you mean, implement these layers and try with the current Tacotron model or try them in isolation?

**Comment by erogol** [comment]
are you on matrix already?  Maybe we can continue our talk there and share updates here.

**Comment by WeberJulian** [comment]
Yeah I meant with the current implementation of Tacotron 2 and instead of our current speaker embeddings. No I'm not on matrix yet, but sure I'll join, it's better than spamming this issue.

**Comment by erogol** [comment]
[Here ](https://matrix.to/#/!NxVmsWBjMEyIHTQCOW:matrix.org?via=matrix.org)is the matrix link to join in our discussion on this.
